src/stdlib_linalg_schur.fypp

```diff
@@ -0,0 +1,358 @@
+            
+        case default
+            
+            err = linalg_state_type(this, LINALG_INTERNAL_ERROR, 'GEES catastrophic error: info=', info)
+
+        end select
+        
+    end subroutine handle_gees_info
+
+    #:for rk, rt, ri in RC_KINDS_TYPES    
+    !> Workspace query
+    module subroutine get_schur_${ri}$_workspace(a,lwork,err)
+        !> Input matrix a[m,m]
+        ${rt}$, intent(in), target :: a(:,:)
+        !> Minimum workspace size for the decomposition operation
+        integer(ilp), intent(out) :: lwork
+        !> State return flag. Returns an error if the query failed
+        type(linalg_state_type), optional, intent(out) :: err
+        
+        integer(ilp) :: m,n,sdim,info
+        character :: jobvs,sort
+        logical(lk) :: bwork_dummy(1)
+        ${rt}$, pointer :: amat(:,:)
+        real(${rk}$) :: rwork_dummy(1)
+        ${rt}$ :: wr_dummy(1),wi_dummy(1),vs_dummy(1,1),work_dummy(1)
+        type(linalg_state_type) :: err0
+        
+        !> Initialize problem
+        lwork = -1_ilp
+        m = size(a,1,kind=ilp)
+        n = size(a,2,kind=ilp)         
+        
+        !> Create a dummy intent(inout) argument
+        amat => a
+        
+        !> Select dummy task
+        jobvs = gees_vectors(.true.)
+        sort  = gees_sort_eigs(.false.)
+        sdim  = 0_ilp
+        
+        !> Get Schur workspace 
+        call gees(jobvs,sort,do_not_select,n,amat,m,sdim,wr_dummy,#{if rt.startswith('r')}#wi_dummy, #{endif}#&
+                  vs_dummy,m,work_dummy,lwork,#{if rt.startswith('c')}#rwork_dummy,#{endif}#bwork_dummy,info)
+        if (info==0) lwork = nint(real(work_dummy(1),kind=${rk}$),kind=ilp)
+        call handle_gees_info(info,m,n,m,err0)
+        call linalg_error_handling(err0,err)
+        
+        contains
+        
+            ! Interface to dummy select routine
+            pure logical(lk) function do_not_select(alpha#{if rt.startswith('r')}#r,alphai#{endif}#) 
+                ${rt}$, intent(in) :: alpha#{if rt.startswith('r')}#r,alphai#{endif}# 
+                do_not_select = .false.
+            end function do_not_select 
+        
+    end subroutine get_schur_${ri}$_workspace   
+
+    ! Schur decomposition subroutine
+    module subroutine stdlib_linalg_${ri}$_schur(a,t,z,eigvals,overwrite_a,storage,err)
+        !> Input matrix a[m,m]
+        ${rt}$, intent(inout), target :: a(:,:)
+        !> Schur form of A: upper-triangular or quasi-upper-triangular matrix T
+        ${rt}$, intent(out), contiguous, target :: t(:,:)
+        !> Unitary/orthonormal transformation matrix Z
+        ${rt}$, optional, intent(out), contiguous, target :: z(:,:)
+        !> [optional] Output eigenvalues that appear on the diagonal of T
+        complex(${rk}$), optional, intent(out), contiguous, target :: eigvals(:)
+        !> [optional] Provide pre-allocated workspace, size to be checked with schur_space
+        ${rt}$, optional, intent(out), target :: storage(:)
```

Comment:
I'm a bit confused by `storage` being `intent(out)`. Given that it is an optional that can be used to provide working storage for the solver, shouldn't it be `intent(inout)` ?

Reply:
[31c43e6](https://github.com/fortran-lang/stdlib/pull/892/commits/31c43e64a4aa30664c11ddcb03063239f6cd4d1f)